### ISSUE TITLE
New databases not created for plugins

### ISSUE THREAD

**Issue by flathat** [CORE-POS/IS4C#479]
I'm testing the enablement of a plugin that has worked before. It uses its own database. After disabling the plugin I set aside the database that it had been using. On re-enabling, in `my_plugin->settingChange()`, `$dbc = FannieDB::get($db_name);` does not create the database if it doesn't exist, as I think it used to. Log shows:
```
(debug) Failed Query on /IS4C/fannie/install/InstallPluginsPage.php
(debug) use my_plugin_database
(debug) Unknown database 'my_plugin_database'
```
Code git merged from upstream yesterday.
The tables for the plugin were created in db `core_calendar` instead.  The `CoreCalendar` plugin, which uses this, is just before mine on the Install > Plugins page.  I disabled the `CoreCalendar` plugin and set aside `core_calendar`. The next enablement of my plugin put the tables in `core_op`.  Re-enabling `CoreCalendar` got a similar error and put the tables in `core_op`.

I created a database for `my_plugin` and re-enabled it and the tables were put there.
`FannieDB.php`
```php
    public static function get($db_name, &$previous_db=null)
    {
        if (!self::dbIsConfigured()) {
            return false;
        } else if (self::$db == null) {
            $previous_db = $db_name;
            self::newDB($db_name);
        } else if (!isset(self::$db->connections[$db_name])) {
            $previous_db = self::$db->defaultDatabase();
            self::addDB($db_name);
        } else {
            $previous_db = self::$db->defaultDatabase();
        }

        self::$db->setDefaultDB($db_name);

        return self::$db;
    }

    private static function dbIsConfigured()
    {
        $config = FannieConfig::factory();
        if ($config->get('SERVER') == '' || $config->get('SERVER_DBMS') == '' || 
$config->get('SERVER_USER') == '') {
            return false;
        } else {
            return true;
        }
    }

    private static function newDB($db_name)
    {
        $config = FannieConfig::factory();
        if (!class_exists('SQLManager')) {
            include(dirname(__FILE__) . '/../../src/SQLManager.php');
        }
        self::$db = new SQLManager(
            $config->get('SERVER'),
            $config->get('SERVER_DBMS'),
            $db_name,
            $config->get('SERVER_USER'),
            $config->get('SERVER_PW'),
            false,
            true);
    }

    private static function addDB($db_name)
    {
        self::$db->selectDB($db_name);
    }

```
An earlier version of `FannieDB->addDB()` calls `add_connection()`:
```php
    private static function addDB($db_name)
    {
        include(dirname(__FILE__).'/../../config.php');
        if (!class_exists('SQLManager')) {
            include($FANNIE_ROOT.'src/SQLManager.php');
        }
        self::$db->add_connection($FANNIE_SERVER,$FANNIE_SERVER_DBMS,
            $db_name, $FANNIE_SERVER_USER, $FANNIE_SERVER_PW);
    }
```
`SQLManager.php`
```php
    /**
      Change the default database on a given connection
      (i.e., mysql_select_db equivalent)
      @param $db_name [string] database name
      @param $which_connection [optional]
      @return boolean

      Using this method will recycle an existing connection
      object where as calling addConnection will create a
      new connection object.
    */
    public function selectDB($db_name, $which_connection='')
    {
        if ($which_connection == '') {
            $which_connection=$this->default_db;
        }

        $current_db = $this->defaultDatabase($which_connection);
        if ($current_db === false) {
            // no connection; cannot switch database
            return false;
        }

        $this->connections[$db_name] = $this->connections[$which_connection];

        return $this->setDefaultDB($db_name);
    }

```

```php
    public function addConnection($server,$type,$database,$username,$password='',
                           $persistent=false,$new=false)
    {
        if (empty($type)) {
            return false;
        }

        $conn = ADONewConnection($type);
        $conn->SetFetchMode(ADODB_FETCH_BOTH);
        $ok = false;
        if (isset($this->connections[$database]) || $new) {
            $ok = $conn->NConnect($server,$username,$password,$database);
        } else {
            if ($persistent) {
                $ok = $conn->PConnect($server,$username,$password,$database);
            } else {
                $ok = $conn->Connect($server,$username,$password,$database);
            }
        }
        $this->connections[$database] = $conn;

        if (!$ok) {
            $conn = ADONewConnection($type);
            $conn->SetFetchMode(ADODB_FETCH_BOTH);
            $ok = $conn->Connect($server,$username,$password);
            if ($ok) {
                $stillok = $conn->Execute("CREATE DATABASE $database");
                if (!$stillok) {
                    $this->connections[$database] = false;
                    return false;
                }
                $conn->Execute("USE $database");
                $this->connections[$database] = $conn;
            } else {
                $this->connections[$database] = false;
                return false;
            }
        }

        return true;
    }
```

**Comment by gohanman** [comment]
That's definitely a bug. SQLManager::setDefaultDB needs to create databases. The basic change (6a19d90aeab0db0c367f0afc4f109942660bb2c8) should be preserved though. Creating multiple separate connections causes a number of different problems. In particular, the life span of connections, re-use of underlying connections, and default database (e.g., "use core_op") has different behavior across mysql/mysqli/PDO. Using exactly one connection is much more consistent.

Now that you bring it up though, there's no guarantee that the DB user account has permission to create additional databases. In that case there should be an error/exception. Otherwise it would be the same behavior: fail to create new database, create tables/views in whatever database is currently in use.

**Comment by flathat** [comment]
The default to create a db if it doesn't exist is probably the source of some odd ones that have appeared on my system from time to time.  The feature is important for enabling plugins but I wonder if it should not be the default, thus:
In `FannieDB.php`:
```php
public static function get($db_name, &$previous_db=null, $create_db=false) {
    //...
    self::$db->setDefaultDB($db_name, $create_db);    
```
In the `FanniePlugin` extension:
```php
  public function settingChange () {
        // ...
        // Creates the database if it doesn't already exist.
        $dbc = FannieDB::get($db_name, NULL, true);
```

**Comment by gohanman** [comment]
I don't think the extra parameter is necessary. I'm having trouble thinking of a time you wouldn't want to create it. Especially if all the places where database names are configurable pass $create_db=true. The only edge case would be if someone hardcoded a database name.

**Comment by gohanman** [comment]
Whoops. I forgot to link to the actual fix: e58aa83d5f28f1d567e0ba21a83a9c9053d56fa4

That will create a database if possible which is consistent with the previous behavior. If creating the database fails, the calling code has to account for it somehow which is a change in behavior. Most importantly, requesting a new database which fails followed by requesting an existing database that succeeds should return a valid connection and that may be a little tricky.

**Comment by gohanman** [comment]
Fix is in master now.
